hudi-cli/src/main/java/org/apache/hudi/cli/commands/DiffCommand.java

```diff
@@ -0,0 +1,196 @@
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.table.timeline.HoodieActiveTimeline;
+import org.apache.hudi.common.table.timeline.HoodieArchivedTimeline;
+import org.apache.hudi.common.table.timeline.HoodieDefaultTimeline;
+import org.apache.hudi.common.table.timeline.HoodieInstant;
+import org.apache.hudi.common.util.NumericUtils;
+import org.apache.hudi.common.util.StringUtils;
+
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.springframework.shell.core.CommandMarker;
+import org.springframework.shell.core.annotation.CliCommand;
+import org.springframework.shell.core.annotation.CliOption;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.apache.hudi.cli.utils.TimelineUtil.getTimeDaysAgo;
+
+/**
+ * Given a file id or partition value, this command line utility tracks the changes to the file group or partition across range of commits.
+ * Usage: diff file --fileId <fileId>
+ */
+@Component
+public class DiffCommand implements CommandMarker {
+
+  private static final Logger LOG = LogManager.getLogger(DiffCommand.class);
+
+  @CliCommand(value = "diff file", help = "Check how file differs across range of commits")
+  public String diffFile(
+      @CliOption(key = {"fileId"}, help = "File ID to diff across range of commits", mandatory = true) String fileId,
+      @CliOption(key = {"startTs"}, help = "start time for compactions, default: now - 10 days") String startTs,
+      @CliOption(key = {"endTs"}, help = "end time for compactions, default: now - 1 day") String endTs,
+      @CliOption(key = {"limit"}, help = "Limit compactions", unspecifiedDefaultValue = "-1") final Integer limit,
+      @CliOption(key = {"sortBy"}, help = "Sorting Field", unspecifiedDefaultValue = "") final String sortByField,
+      @CliOption(key = {"desc"}, help = "Ordering", unspecifiedDefaultValue = "false") final boolean descending,
+      @CliOption(key = {"headeronly"}, help = "Print Header Only", unspecifiedDefaultValue = "false") final boolean headerOnly,
+      @CliOption(key = {"includeArchivedTimeline"}, help = "Include archived commits as well",
+          unspecifiedDefaultValue = "false") final boolean includeArchivedTimeline) throws IOException {
+    HoodieDefaultTimeline timeline = getTimelineInRange(startTs, endTs, includeArchivedTimeline);
+    return printCommitsWithMetadataForFileId(timeline, limit, sortByField, descending, headerOnly, "", fileId);
+  }
+
+  @CliCommand(value = "diff partition", help = "Check how file differs across range of commits")
+  public String diffPartition(
+      @CliOption(key = {"partitionPath"}, help = "Relative partition path to diff across range of commits", mandatory = true) String partitionPath,
```

Comment:
what value should we set it for non-partitioned tables?

Reply:
This command tracks changes to a partition. It is meant to be used only for partitioned tables. I'll make it clear in the `help`. Without specifying `--partitionPath`, this command will fail.